src/components/LocalFileUpload/custom-request-upload.ts

```diff
@@ -1,39 +1,113 @@
+        // this function is called multiple times, but we only need to process
+        // and load the trajectory once
+        return;
+    }
+
     // want the loading indicator to show without any lag time
     // as soon as user hits "Open" button, and not have to have this action called
     // multiple places in the code.
     store.dispatch({
         payload: { newFile: true },
         type: CLEAR_SIMULARIUM_FILE,
     });
-    file.text()
-        .then((text) => JSON.parse(text))
-        .then((data) => {
-            loadFunction({
-                lastModified: file.lastModified,
-                name: file.name,
-                data,
-            });
-        })
-        .then(() =>
-            onSuccess(
-                {
-                    name: file.name,
-                    status: "done",
-                    url: "",
-                },
-                file
-            )
-        )
-        .catch((error) => {
-            console.log(error);
-            onError(error);
-        });
+
+    const files: FileHTML[] = Array.from(droppedFiles) as FileHTML[];
+
+    Promise.all(files.map((file: FileHTML) => file.text())).then(
+        (parsedFiles: string[]) => {
+            try {
+                const simulariumFileIndex = findIndex(files, (file) =>
+                    file.name.includes(".simularium")
+                );
+                // TODO: handle when user doesn't load a .simularium file
+                // if (simulariumFileIndex === -1) {
+                //    throw new Error("Please upload a .simularium file.")
+                // }
+
+                const simulariumFile: File = files[simulariumFileIndex];
+                const simulariumFileJson: SimulariumFileFormat = JSON.parse(
+                    parsedFiles[simulariumFileIndex]
+                );
+                const geoAssets: { [name: string]: string } = files.reduce(
+                    (acc: { [name: string]: string }, cur, index) => {
+                        if (index !== simulariumFileIndex) {
+                            acc[cur.name] = parsedFiles[index];
+                        }
+                        return acc;
+                    },
+                    {}
+                );
+
+                loadFunction({
+                    lastModified: simulariumFile.lastModified,
+                    name: simulariumFile.name,
+                    data: simulariumFileJson,
+                    geoAssets: geoAssets,
+                });
+                // TS thinks onSuccess might be undefined
+                if (onSuccess) {
+                    onSuccess(
+                        {
+                            name: simulariumFile.name,
+                            status: "done",
+                            url: "",
+                        },
+                        new XMLHttpRequest() // onSuccess needs an XMLHttpRequest arg
+                    );
+                }
+                isLoading = false;
```

Comment:
That is a really good point. I'll make this change

Reply:
yeah this seems good, but it does mean you're making the user wait *slightly longer, I was thinking you'd process on the first call but not reset anything, and then reset the counter on the last call? I am not sure I understand why that would cause a problem

Reply:
Oh my god why didn't I think of that.... hang on :)

Reply:
Done